### ISSUE TITLE
AV when pasting and when replacing texts (with fix)

### ISSUE THREAD

**Issue by ClementDoss** [TurboPack/SynEdit#94]
Hi
This is my first bug report.
I'm using Synedit mostly for SQL syntax and resolved AV which happened when copy/pasting or searching/replacing texts.

The first fix is in PasteFromClipboard method line 3376:
From
  // SetSelTextPrimitiveEx Encloses the undo actions in Begin/EndUndoBlock
  SetSelTextPrimitiveEx(PasteMode, GetClipboardText, True);

to

    **Inc(FPaintTransientLock);**  // added this line
    SetSelTextPrimitiveEx(PasteMode,GetClipboardText, True);
    **Dec(FPaintTransientLock);**  // added this line

The other one is in SearchReplace method line 7250
**from:**
 try
    while (ptCurrent.Line >= ptStart.Line) and (ptCurrent.Line <= ptEnd.Line) do
    begin
      nInLine := fSearchEngine.FindAll(Lines[ptCurrent.Line - 1]);

**to:**
**Inc(FPaintTransientLock);**  // added this line
 try
    while (ptCurrent.Line >= ptStart.Line) and (ptCurrent.Line <= ptEnd.Line) do
    begin
      nInLine := fSearchEngine.FindAll(Lines[ptCurrent.Line - 1]);

**and at 7344:**

**from:**
  finally
    if bReplaceAll and not bPrompt then DecPaintLock;
    if bEndUndoBlock then EndUndoBlock;
    DoOnPaintTransient( ttAfter );
  end;

**to**

  finally
    if bReplaceAll and not bPrompt then DecPaintLock;
    if bEndUndoBlock then EndUndoBlock;
   **Dec(FPaintTransientLock);**  // added this line
    DoOnPaintTransient( ttAfter );
  end;

With those fixes the AV stop happening, but I know nothing about this code base and I'm not sure this is the right way to fix it.
HTH,
Clément

**Comment by pyscripter** [comment]
Thanks for the report.  Please note that:

1. You are not using the current version from this repo.  Could you please uninstall your version and **install the one from this repo**? Clone or download and the install.
2. You should **provide a simple project that reproduces this error**.  Do you get the error(s) with the Search and Replace demo of this repo.
3. This is not the right way to fix it. You should not manipulate FPaintTransientLock directly.

For the search and replace demo try the following.  In TCustomSynEdit.SearchReplace:

Replace  the first DoOnPaintTransient with
          DoOnPaintTransientEx(ttBefore,true);
and the second with
          DoOnPaintTransientEx(ttAfter,true);

Does this solve your problem?

Regarding the PasteFromClipboard you could the same, but I cannot see how this may have an impact.  PasteFromClipboard calls SetSelTextPrimitiveEx which does not call DoOnPaintTransient.  At least not in the code of this repo.  Using the version from this repo should not have such a problem.

And in any case what exactly are you doing in your handling of OnPaintTransient that causes the error?

**Comment by ClementDoss** [comment]
Usually the AV happened when copy/pasting over 2000 lines, when inserting at cursor position over 2000 lines or when search replacing in those 2000 lines.
I have updated my version with the latest version, and it seems the AV is gone. I didn't make the changes you suggested. 
In the event OnPaintTransient I'm highlighting matching '(', '{', '[' with their corresponding ')','}',']'.

I still have not manage to duplicate the AV with this latest version. And I tested with 5000 lines texts.
I'll keep trying, and report if I manage to duplicate.

thanks a lot
